### PR TITLE
LOKI Milestone Report

### PR DESCRIPTION
**Pull request by LamarMoore**
Report to assess Mantid performance for live reduction of LOKI data.

### CODE REVIEW

**Review by DanNixon — APPROVED** [review]

File: investigations/LOKI/loki_performance.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
+# LOKI Live Reduction Performance in Mantid
+
+## Motivation
+
+The LOKI instrument for small angle neutron scattering is scheduled to be one of the first instruments to come online for the ESS. As such, it is important that the day 1 requirements for the software are satisfied before the instrument enters the commissioning phase. LOKI will be ~1Mpixels with an incident flux between 10<sup>5</sup>Hz and 10<sup>8</sup>Hz (worst case). This represents a significant processing and storage challenge for the ESS. Since Mantid has been chosen as the data-reduction platform at the ESS, it is critical that the software can cope with these data rates. This document outlines the current performance of the Mantid framework and recommendations for improvements which may allow performance to meet the requirements.
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md]: *When you say cope with rates, this is not qualified in the context of needing to perform live in-situ data reduction. That is also part of the DMSC remit and should be referenced.
*The report needs to include references to the dates and rates it provides.
*Relevant links for LOKI (instrument itself) would also be useful.

File: investigations/LOKI/loki_performance.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+## Scope
+This document only covers performance benchmarks for the LOKI instrument based on the geometry described in the proceeding section. A general benchmarking exercise of the live data reduction in Mantid was performed [here](https://github.com/DMSC-Instrument-Data/documents/blob/master/investigations/Live%20Reduction/LiveReductionInvestigation.md) and this document may be referenced within.
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md]: Isn't the scope set by the High Level milestones? i.e. Transition to Operations?

File: investigations/LOKI/loki_performance.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+## LOKI Geometry
+The LOKI Geometry used for these tests are as specified in the engineering diagrams below. The instrument consists of 864 tubes (6048 straws) with an intrinsic resolution of 5mm. Straw lengths vary between 1.2m and 0.5m which gives a total pixel count of 1,245,880pixels.
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md]: Minor point, but maybe a sentence on this design being quite recent. The Band GEM route being discarded in 2018/2017?

File: investigations/LOKI/loki_performance.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+![geometry](loki_geometry.png)
+![mantid](loki_mantid.png)
+
+## Build-out Phases
+
+Currently "Day 1" operations for LOKI are set to commence November 2022. The initial instrument will contain the rear detector bank and 2 panels (one horizontal and one vertical) in the front and middle banks each. The resulting instrument will consist of ~473K pixels. However based on initial simulations, most of the detected neutron beam will be concentrated on the rear bank. The remaining banks are currently set to be added to the final instrument between 1-5 years beyond 2022. Hot commissioning and early science for this instrument are set for mid 2023 with full operations at the start of 2025.
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md:67]: 👍 Really good. See my point above regarding any reference if possible.

File: investigations/LOKI/loki_performance.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+## Hardware 
+
+The hardware used for these tests is as follows:
+
+- 128GB RAM
+- Intel(R) Core(TM) i9-9920X CPU @ 3.50GHz
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md:74]: Is the OS relevant?

File: investigations/LOKI/loki_performance.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+### Previous Efforts
+The performance of Mantid for live data streaming/reduction was evaluated and results presented in the following document [LiveReductionInvestigation](https://github.com/DMSC-Instrument-Data/documents/blob/master/investigations/Live%20Reduction/LiveReductionInvestigation.md). Following from this work Lamar Moore (ISIS), Dan Nixon (ISIS/ESS-in-kind) and Simon Heybrock (ESS) met December 2018 in Copenhagen to discuss and prototype strategies for addressing performance bottlenecks in this area. The result of this work can be found [here](https://github.com/DMSC-Instrument-Data/documents/blob/master/meeting_notes/December_2018/agenda.md). After this meeting the team reached the following conclusions:
+
+- Optimizing Mantid to satisfy ESS performance requirements is achievable with current time and resource.
+- The optimization effort can be carried out without the need for MPI (distribution of the event stream for use on a HPC Cluster).
+- The optimizations are straightforward.
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md]: straight forward in terms of what?

File: investigations/LOKI/loki_performance.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+The performance of Mantid for live data streaming/reduction was evaluated and results presented in the following document [LiveReductionInvestigation](https://github.com/DMSC-Instrument-Data/documents/blob/master/investigations/Live%20Reduction/LiveReductionInvestigation.md). Following from this work Lamar Moore (ISIS), Dan Nixon (ISIS/ESS-in-kind) and Simon Heybrock (ESS) met December 2018 in Copenhagen to discuss and prototype strategies for addressing performance bottlenecks in this area. The result of this work can be found [here](https://github.com/DMSC-Instrument-Data/documents/blob/master/meeting_notes/December_2018/agenda.md). After this meeting the team reached the following conclusions:
+
+- Optimizing Mantid to satisfy ESS performance requirements is achievable with current time and resource.
+- The optimization effort can be carried out without the need for MPI (distribution of the event stream for use on a HPC Cluster).
+- The optimizations are straightforward. 
+
+**There were previous small optimizations made as part of the live reduction investigation which resulted in a small performance improvement. Tests conducted in this investigation are from that basis.**
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md]: Not 100% clear on what "from that basis" means here. Maybe reword.

File: investigations/LOKI/loki_performance.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+SANSSolidAngleCorrection(InputWorkspace="ws", OutputWorkspace="ws", DetectorTubes=False, DetectorWIng=False)
+SANSAbsoluteScale(InputWorkspace="ws", OutputWorkspace="ws", ScalingFactor=0.99)
+Rebin(InputWorkspace="ws", OutputWorkspace="ws", Params="74294664", PreserveEvents=False)
+Q1D(DetBankWorkspace="ws", OutputWorkspace="SANSReduced", SolidAngleWeighting=False, OutputBinning="1,5,100")
+```
+
+Data consumption rates averaged ~18Hz (messages) at an event rate of ~1.4e<sup>7</sup>Hz. All other configurations were fixed as in the previous case.
+
+### Summary and Final Conclusions
+The final optimizations added to Mantid have facilitated live streaming which will be exceedingly capable of coping with projected instrument flux. The scalability of the solution also demonstrates that as processor architectures improve and scale in future, iterations, we will be in a good position to manage increasing flux. There are currently no recommended actions required for parallelisation/distribution of Mantid for live streaming.
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md]: Minor point "exceedingly capable" Other than that this is good/

File: investigations/LOKI/loki_performance.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+#### Writing To Workspace
+
+Data consumption rates averaged ~24Hz (messages) at an event rate of ~1.4e<sup>7</sup>Hz. This utilized a 30s timeout for `MonitorLiveData` and a basic `Rebin` operation for displaying data on the instrument. The event buffer stored 2.5e<sup>7</sup> events before writing to the workspace.
+
+##### Scaling
+There was previous work done in investigating how well the live streamer in Mantid scales and was presented in this document [LiveReductionInvestigation](https://github.com/DMSC-Instrument-Data/documents/blob/master/investigations/Live%20Reduction/LiveReductionInvestigation.md). Scaling for LOKI ingesting the simulated data showed the following trend:
+
+`#` of threads|Message Consumption Rate
+---|---
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md]: This is very important, but I'm guessing this is going to be obtained before the report is finalised.

File: investigations/LOKI/loki_performance.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+## Scope
+This document only covers performance benchmarks for the LOKI instrument based on the geometry described in the proceeding section. A general benchmarking exercise of the live data reduction in Mantid was performed [here](https://github.com/DMSC-Instrument-Data/documents/blob/master/investigations/Live%20Reduction/LiveReductionInvestigation.md) and this document may be referenced within.
+
+## LOKI Geometry
+The LOKI Geometry used for these tests are as specified in the engineering diagrams below. The instrument consists of 864 tubes (6048 straws) with an intrinsic resolution of 5mm. Straw lengths vary between 1.2m and 0.5m which gives a total pixel count of 1,245,880pixels.
+![geometry](loki_geometry.png)
+![mantid](loki_mantid.png)
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md:63]: Unfortunately you can't see much from the instrument view. Maybe enable lighting and change colour would show the tubes?

**Review by OwenArnold — CHANGES_REQUESTED** [review]
I think you have probably now close to completion with this, but the report reads the wrong way round to me. The most important things that should be addressed are:

* How does the software now satisfy the 2019 requirements, which you do not reference, but are the fundamental reason we did this work [see here](https://confluence.esss.lu.se/pages/viewpage.action?pageId=262411283)
* Ingest + Scalability of ingest
* DR itself (report output) with comments on recommendations, which you already mention which is great. You should state that we are using the ISIS SANS workflow, which is the agreed best analogue of what will happen at the ESS. Scalability of DR, which you also allude to. And how we might tackle these issues between now and SOUP.
* Mentions of further demonstration for Ingest  V20 tests September 2019, (though these are not conditional for passing the milestone) 

I would extract all the technical details to latter sections. The reader should not have to read through the whole document to get to the conclusions. Thomas Rod will not read or necessarily understand what is in it at present.

File: investigations/LOKI/loki_performance.md (hunk 5 above)
Comment (LamarMoore) [reply investigations/LOKI/loki_performance.md:74]: Don't think so.

File: investigations/LOKI/loki_performance.md (hunk 6 above)
Comment (LamarMoore) [reply investigations/LOKI/loki_performance.md]: I will add a note "in comparison to other development efforts. For example, SciPP and NexusGeometry efforts where the problems are comparatively very complex and require large amounts of effort."

File: investigations/LOKI/loki_performance.md (hunk 5 above)
Comment (DanNixon) [reply investigations/LOKI/loki_performance.md:74]: Might be worth mentioning OpenMP support must be present. We don't have this on Mac, not sure about Windows.

However, since the ESS are set on using CentOS it may not be important enough to mention.

File: investigations/LOKI/loki_performance.md (hunk 6 above)
Comment (OwenArnold) [reply investigations/LOKI/loki_performance.md]: I think just to say that this was predicted to be 1-2 months (insert as appropriate) for a single experienced developer.

File: investigations/LOKI/loki_performance.md (hunk 7 above)
Comment (LamarMoore) [reply investigations/LOKI/loki_performance.md]: As in, these tests were performed after a few optimizations were put in place. As opposed to Mantid in it's current state in the present release.

File: investigations/LOKI/loki_performance.md (hunk 9 above)
Comment (LamarMoore) [reply investigations/LOKI/loki_performance.md]: Yes @DanNixon  is getting these figures

File: investigations/LOKI/loki_performance.md (hunk 10 above)
Comment (LamarMoore) [reply investigations/LOKI/loki_performance.md:63]: I can do profiles/zoomed shots so the tubes and straws are clearly visible.

Comment (OwenArnold) [reply investigations/LOKI/loki_performance.md:63]: I know that the geometry was extremely hard to create, unfortunately the no-data visualisation doesn't do it justice.

File: investigations/LOKI/loki_performance.md (hunk 2 above)
Comment (LamarMoore) [reply investigations/LOKI/loki_performance.md]: I thought that was the motivation. I thought the scope was determined by the desired use-case i.e. live data reduction.

**Comment by DanNixon** [comment]
Thinking about it the loss of performance at 24 threads may also be physical core oversubscription due to HyperThreading.

File: investigations/LOKI/loki_performance.md (hunk 9 above)
Comment (LamarMoore) [reply investigations/LOKI/loki_performance.md]: @DanNixon Should these numbers be divided by the number of banks?

Comment (DanNixon) [reply investigations/LOKI/loki_performance.md]: Maybe it is easier to present it that way. I left it as messages since that's what the header stated and this measurement is less instrument specific.

Yes, dividing by number of banks (9) will give the maximum pulse rate which is probably a clearer indication of whether or not LOKI can be streamed.

Comment (LamarMoore) [reply investigations/LOKI/loki_performance.md]: Ok I will do that as I am making changes to the document. Thanks Dan

File: investigations/LOKI/loki_performance.md [hunk 11]
```diff
@@ -24,12 +25,41 @@ The hardware used for these tests is as follows:
...
+
+`#` of threads|Message Consumption Rate (Hz)
+---|---
+1|8
+4|26
+8|33
+16|52
+24 (Hyperthreading)|21
+
+Note that at 24 threads it is assumed the host system is oversubscribed due to running Mantid, Kafka broker and NeXus publisher simultaneously as well as hyperthreading.
```
Comment (DanNixon) [inline investigations/LOKI/loki_performance.md]: I might have not been overly clear, my point regarding HyperThreading was that loading all logical threads with a relatively constant workload is typically worse than loading one logical thread per physical core (due to context switching, etc.).

HyperThreading was always enabled, but at 24 logical threads these effects could have started to become visible. Although the sorting and insertion are fairly quick to execute so still think thread oversubscription was the issue.

Comment (LamarMoore) [reply investigations/LOKI/loki_performance.md]: Feel free to make the relevant change to the document

File: investigations/LOKI/loki_performance.md [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+
+`#` of threads|Message Consumption Rate (Hz)
+---|---
+1|8
+4|26
+8|33
+16|52
+24|21
+
+Note that at 24 threads it is assumed the host system is oversubscribed due to running Mantid, Kafka broker and NeXus publisher simultaneously as well as hyperthreading.
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md]: 👍 A plot might be nice.

File: investigations/LOKI/loki_performance.md [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+- Target Linux Platform (OpenMP is a hard requirement)
+
+## Outcomes
+
+### Measurement Interpretation
+The ESS pulse rate for neutrons is expected to be **14Hz**. This will be the rate at which messaged will be produced in the Kafka system and therefore consumed by Mantid. Measurements are taken in this context. If Mantid is capable of processing events (writing events to in-memory data structures and performing basic reduction) at this rate at a target neutron flux, this indicates that Mantid can cope with the event rate.
+
+### Data ingest into Mantid
+
+Data consumption into Mantid using realistic SANS data generated using GEANT4 resulted in an average consumption rate of ~24Hz. This was performed at an event flux of ~1.4e<sup>7</sup> events per message. This clearly shows that Mantid can comfortably deal with the ingest of data at these high rates.
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md:84]: 10^8 was also mentioned. Was this measured, can anything be said about this? I don't we should worry if not as the Milestone did not mention it.

File: investigations/LOKI/loki_performance.md [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+![Sans Reduction Profile](sansreductionprofile.png)
+
+In this test, 12 CPU cores were allocated to the data reduction, we see the reduction is mostly using all allocated cores most of the time. This suggests this workflow is currently optimised to make use of parallel CPU architectures. However, `ConvertUnits` seems to be entirely single threaded and half of the `Rebin` operation is single-threaded. Recommendations for optimizing the workflow are presented in the last section of this document.
+
+### Ingest with ISIS SANS Workflow
+
+The ingest of data into Mantid was coupled with the ISIS SANS Workflow to assess the impact on the message consumption rate. Mantid reported an average consumption rate of ~18Hz. This is still above the 14Hz target but there is the possibility of optimizing the SANS workflow algorithms to obtain better performance.
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md:110]: Great. This is one of the most important outcomes of the report as far as the readers will be concerned.

File: investigations/LOKI/loki_performance.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
+# LOKI Live Reduction Performance in Mantid
+
+## Motivation
+
+The [LOKI instrument](https://europeanspallationsource.se/instruments/loki) for small angle neutron scattering is scheduled to be one of the first instruments to come online for the ESS. As such, one of the [High Level DMSC Milestones](https://confluence.esss.lu.se/pages/viewpage.action?pageId=262411283) is to ensure the data reduction software can cope with the proposed instrument flux in a live data reduction scenario. LOKI will be ~1Mpixels with an incident flux between 10<sup>5</sup>Hz and 10<sup>8</sup>Hz (worst case). This represents a significant processing and storage challenge for the ESS. Since Mantid has been chosen as the data-reduction platform at the ESS, it is critical that the software can cope with the expected data rates in preparation for hot comissioning and early science in mid 2023. The maximum instrument flux during these phases is set to be ~10<sup>7</sup>Hz. This document outlines the current performance of the Mantid framework and recommendations for improvements which may allow performance to meet the requirements.
```
Comment (OwenArnold) [inline investigations/LOKI/loki_performance.md]: I think we are probably not being quite explicit enough about which milestone and how it has been met. I'm happy to help with this.

File: investigations/LOKI/loki_performance.md (hunk 13 above)
Comment (OwenArnold) [reply investigations/LOKI/loki_performance.md:84]: Along these lines - how can we re-run all facets of the tests/measurements required for this report. It might be good to give links to all branches/repos or maybe there is an easier way? I don't think this should delay this report at all though.

Comment (LamarMoore) [reply investigations/LOKI/loki_performance.md:84]: No I mentioned this in the scope. We limited the measurements to ~10e7 in line with the projected rates during Hot Commissioning/early science.

Comment (LamarMoore) [reply investigations/LOKI/loki_performance.md:84]: @OwenArnold  The links to the various repos are within the document (e.g the kafka streamer). The only thing which is missing is Mantid, however all of the diagnostics are intrusive so we wouldn't have a release version of Mantid which will spit out benchmarks.

**Review by OwenArnold — APPROVED** [review]
